readme.txt
For installation instructions, please read doc/INSTALL.


doc/CHANGELOG
$Id: CHANGELOG 3810 2007-12-05 19:13:25Z brian $

CATS ChangeLog
    _____________________________________________________________________

                               CATS ChangeLog

12/05/07
  * BH: Various changes to backup process and graph statistics for
        compatibility with older versions of MySQL.
  * BH: TinyMCE added to edit job order descriptions and notes.
  * BH: Added CATS Professional user class - Career Portal Customizer
  * BH: Added Job Order Datagrid column - Number of Not Contacted Pipeline

11/26/07
  * BH: Saved list management removed from edit candidate.
  * BH: Added field best time to call to candidates.
  * BH: PHP 4 doesn't generate a PHP error, but says PHP is out of
        date in installer instead.

11/22/07
  * BH: CATS checks PHP version when index.php loads now.  This prevents
        PHP 4 users from seeing a cryptic error message because they
        don't have the required functions for CATS.

11/21/07
  * BH: Accessing CATS from a different directory forces a logout of
        the current session.
  * BH: Minor style changes.

11/20/07
  * BH: List encoding fixed.  Even supports unicode characters!

11/19/07
  * BH: Extra fields are now renameable.

11/14/07
  * BH: Replaced old career portal with new Career portal by Sae.
  * BH: Extra fields are now reorderable.

11/13/07
  * BH: Filecompressor extraction works.
  * BH: Restoring from backup works.
  * BH: Restoring from demo works.

11/09/07
  * BH: Calendar can now send e-mail reminders to multiple recipients.

11/08/07
  * BH: Job Orders now tracks openings available and total openings separately.

11/06/07
  * BH: $_GET, $POST, and $_REQUEST's array keys are also handled by 
        stripslashes if magic quotes are on.  This fixes Saved Lists not 
        sortable with magic quotes on.
  * BH: Placing a candidate reduces number of openings by one.
  * BH: Public calendar entries do not show up on dashboard - just personal ones.
  * BH: Set event duration is added to add activity change status popup.  
        Adding an all day event disabled duration (both on popup and 
        calendar.)

10/31/07
  * BH/WB: Properly handle uppercase file extensions when converting documents.
           Thanks to MePHiSTY and hardik.panchal on the CATS Forums.

10/29/07
  * BH: Users can no longer commit data imports.  They loose the ability to
		revert an import after 1 week.
  * BH: Installer now shows progress while updating the schema.

10/25/07
  * BH: Star ranking can now be set to 0 stars (there is 3 pixels to the
        left of the first star to set 0 rank)
  * BH: Status negotiating is now status qualifying.
  * BH: Fix to datagrid.
  * BH/AK: No data images added throughout CATS.

10/19/07
  * BH: First beta of dashboard - still needs a few gui changes from Asim.

10/12/07
  * WB: Added basic sanity checking of query results to the S3 Storage module.
  * WB: Fixed a bug in the S3 Storage module where the resultset could be lost
        if the file was renamed due to bad characters in the filename.
  * WB: General S3 Storage module cleanups.

09/25/07
  * BH: Removing from saved list via action menu works.
  * BH: Page number less than 0 or more than total pages will constrain users 
        view to within the number of pages in a datagrid.
        
09/24/07
  * BH: Submodal is now a core javascript library, and is included in every file.
        PrintPopupArea is no longer necessary.
  * BH: Multiple items can be added to a list at once.
  * BH: Multiple candidates can be added to a pipeline at once.
  * BH: All popup windows in the system revised to be slimmer and have the title
        in the popup titlebar rather than inline in the popup content.
  * BH: All datagrids in the system now have an item count at the top of the
        datagrid.  (Indicating, for example, number of candidates in the system).
  * BH: Joborders, Companies, and Contacts now have static saved lists.
  * BH: Datagrid modified so a separate instance can be made of each saved list
        rather than shared layout preferences for each saved list.  Saved lists 
        can now be aware of what ID of a list (or other future data item, like a
        saved search) they are rendering.
  * BH: Lists are now added to MRU.
  * BH: Lists are quicksearchable.

09/13/2007
  * BH: ASP Password retrieval fixed.

09/07/2007
  * BH: Attachments are now unauthenticated files delivered through a PHP streamer
        function.

09/06/2007
  * BH: Action area all cleaned up to make it much easier to script in bulk
        actions on items in a datagrid.  For now, only get actions are supported,
        but ajax style popups will soon follow.

08/28/2007
  * BH: Added checkbox for "Only Hot Contacts" and fixed nonfunctional checkbox
        "Only My Contacts".
        
08/24/2007
  * BH: Fixed a horrible bug in attachments where files were not being deleted
        from the server when removed from the database.
  * BH: Added hooks to attachments to allow other modules to interact
         with the act of storing and retrieving attachments.

08/22/2007
  * BH: Various changes.
  * BH: Events can now have reminders set at time of event creation for candidates
        and contacts.

08/21/2007
  * BH: Added a document reindexer that will reindex any resume that has been
        added while a document to text converter for that datatype was unavailable.
        It is invoked at installer time.

08/14/2007
  * BH: Career portal now accepts URL's with unusual characters in JoborderID
        and filters them out.
  * BH: Activity dates/times now editable.
  * BH: Users can now optionally not see EEO information (enabled by default).

07/31/2007
  * BH: Job orders can now be created off of an existing Job Order.

07/25/2007
  * BH: Every E-Mail feature in CATS is now disableable.


index.php
<?php
/*
 * CATS
 * Index (Delegation Module)
 *
 * CATS Version: 0.8.0 (Jhelum)
 *
 * Copyright (C) 2005 - 2007 Cognizo Technologies, Inc.
 *
 *
 * The contents of this file are subject to the CATS Public License
 * Version 1.1a (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.catsone.com/. Software distributed under the License is
 * distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * rights and limitations under the License.
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * The Original Code is "CATS Standard Edition".
 *
 * The Initial Developer of the Original Code is Cognizo Technologies, Inc.
 * Portions created by the Initial Developer are Copyright (C) 2005 - 2007
 * (or from the year in which this file was created to the year 2007) by
 * Cognizo Technologies, Inc. All Rights Reserved.
 *
 *
 * A properly formatted query string will look like this:
 *
 *    /index.php?m=candidates&a=edit&candidateID=55
 *
 *
 * $Id: index.php 3807 2007-12-05 01:47:41Z will $
 */

/* Do we need to run the installer? */
if (!file_exists('INSTALL_BLOCK') && !isset($_POST['performMaintenence']))
{
    include('modules/install/notinstalled.php');
    die();
}

// FIXME: Config file setting.
@ini_set('memory_limit', '64M');

/* Hack to make CATS work with E_STRICT. */
if (function_exists('date_default_timezone_set'))
{
    @date_default_timezone_set(date_default_timezone_get());
}

/* Start error handler if ASP error handler exists and this isn't a localhost
 * connection.
 */
if (file_exists('modules/asp/lib/ErrorHandler.php') &&
    @$_SERVER['REMOTE_ADDR'] !== '127.0.0.1' &&
    @$_SERVER['REMOTE_ADDR'] !== '::1' &&
    substr(@$_SERVER['REMOTE_ADDR'], 0, 3) !== '10.')
{
    include_once('modules/asp/lib/ErrorHandler.php');
    $errorHandler = new ErrorHandler();
}

include_once('./config.php');
include_once('./constants.php');
include_once('./lib/CommonErrors.php');
include_once('./lib/CATSUtility.php');
include_once('./lib/DatabaseConnection.php');
include_once('./lib/Template.php');
include_once('./lib/Users.php');
include_once('./lib/MRU.php');
include_once('./lib/Hooks.php');
include_once('./lib/Session.php'); /* Depends: MRU, Users, DatabaseConnection. */
include_once('./lib/UserInterface.php'); /* Depends: Template, Session. */
include_once('./lib/ModuleUtility.php'); /* Depends: UserInterface */
include_once('./lib/TemplateUtility.php'); /* Depends: ModuleUtility, Hooks */


/* Give the session a unique name to avoid conflicts and start the session. */
@session_name(CATS_SESSION_NAME);
session_start();

/* Try to prevent caching. */
header('Last-Modified: ' . gmdate('D, d M Y H:i:s') . ' GMT');
header('Expires: Mon, 26 Jul 1997 05:00:00 GMT');

/* Make sure we aren't getting screwed over by magic quotes. */
if (get_magic_quotes_runtime())
{
    set_magic_quotes_runtime(0);
}
if (get_magic_quotes_gpc())
{
    include_once('./lib/ArrayUtility.php');

    $_GET     = array_map('stripslashes', $_GET);
    $_POST    = array_map('stripslashes', $_POST);
    $_REQUEST = array_map('stripslashes', $_REQUEST);
    $_GET     = ArrayUtility::arrayMapKeys('stripslashes', $_GET);
    $_POST    = ArrayUtility::arrayMapKeys('stripslashes', $_POST);
    $_REQUEST = ArrayUtility::arrayMapKeys('stripslashes', $_REQUEST);
}

/* Objects can't be stored in the session if session.auto_start is enabled. */
if (ini_get('session.auto_start') !== '0' &&
    ini_get('session.auto_start') !== 'Off')
{
    die('CATS Error: session.auto_start must be set to 0 in php.ini.');
}

/* Proper extensions loaded?! */
if (!function_exists('mysql_connect') || !function_exists('session_start'))
{
    die('CATS Error: All required PHP extensions are not loaded.');
}

/* Make sure we have a Session object stored in the user's session. */
if (!isset($_SESSION['CATS']) || empty($_SESSION['CATS']))
{
    $_SESSION['CATS'] = new CATSSession();
}

/* Start timer for measuring server response time. Displayed in footer. */
$_SESSION['CATS']->startTimer();

/* Check to see if the server went through a SVN update while the session
 * was active.
 */
$_SESSION['CATS']->checkForcedUpdate();


/* We would hook this, but the hooks aren't loaded by the time this code executes.
 * if ASP module exists (code is running on catsone.com), load the website by default
 * rather than the login page.
 */
if (ModuleUtility::moduleExists("asp") && ModuleUtility::moduleExists("website"))
{
    // FIXME: Can we optimize this a bit...?
    include_once('modules/asp/lib/General.php');

    if (!(isset($careerPage) && $careerPage) &&
        !(isset($rssPage) && $rssPage) &&
        !(isset($xmlPage) && $xmlPage) &&
        (!isset($_GET['m']) || empty($_GET['m'])) &&
        (Asp::getSubDomain() == '' || isset($_GET['a'])))
    {
        ModuleUtility::loadModule('website');
        exit(1);
    }
}


/* Check to see if the user level suddenly changed. If the user was changed to disabled,
 * also log the user out.
 */
// FIXME: This is slow!
if ($_SESSION['CATS']->isLoggedIn())
{
    $users = new Users($_SESSION['CATS']->getSiteID());
    $forceLogoutData = $users->getForceLogoutData($_SESSION['CATS']->getUserID());

    if (!empty($forceLogoutData) && ($forceLogoutData['forceLogout'] == 1 ||
        $_SESSION['CATS']->getRealAccessLevel() != $forceLogoutData['accessLevel']))
    {
        $_SESSION['CATS']->setRealAccessLevel($forceLogoutData['accessLevel']);

        if ($forceLogoutData['accessLevel'] == ACCESS_LEVEL_DISABLED ||
            $forceLogoutData['forceLogout'] == 1)
        {
            /* Log the user out. */
            $unixName = $_SESSION['CATS']->getUnixName();

            $_SESSION['CATS']->logout();
            unset($_SESSION['CATS']);
            unset($_SESSION['modules']);

            $URI = 'm=login';

            if (!empty($unixName) && $unixName != 'demo')
            {
                $URI .= '&s=' . $unixName;
            }

            CATSUtility::transferRelativeURI($URI);
            die();
        }
    }
}

/* Check to see if we are supposed to display the career page. */
if (((isset($careerPage) && $careerPage) ||
    (isset($_GET['showCareerPortal']) && $_GET['showCareerPortal'] == '1')))
{
    ModuleUtility::loadModule('careers');
}

/* Check to see if we are supposed to display an rss page. */
else if (isset($rssPage) && $rssPage)
{
    ModuleUtility::loadModule('rss');
}

else if (isset($xmlPage) && $xmlPage)
{
    ModuleUtility::loadModule('xml');
}

/* Check to see if the user was forcibly logged out (logged in from another browser). */
else if ($_SESSION['CATS']->isLoggedIn() &&
    (!isset($_GET['m']) || ModuleUtility::moduleRequiresAuthentication($_GET['m'])) &&
    $_SESSION['CATS']->checkForceLogout())
{
    // FIXME: Unset session / etc.?
    ModuleUtility::loadModule('login');
}

/* If user specified a module, load it; otherwise, load the home module. */
else if (!isset($_GET['m']) || empty($_GET['m']))
{
    if ($_SESSION['CATS']->isLoggedIn())
    {
        $_SESSION['CATS']->logPageView();

        if (!eval(Hooks::get('INDEX_LOAD_HOME'))) return;

        ModuleUtility::loadModule('home');
    }
    else
    {
        ModuleUtility::loadModule('login');
    }
}
else
{
    if ($_GET['m'] == 'logout')
    {
        /* There isn't really a logout module. It's just a few lines. */
        $unixName = $_SESSION['CATS']->getUnixName();

        $_SESSION['CATS']->logout();
        unset($_SESSION['CATS']);
        unset($_SESSION['modules']);

        $URI = 'm=login';
                                 /* Local demo account doesn't relogin. */
        if (!empty($unixName) && $unixName != 'demo')
        {
            $URI .= '&s=' . $unixName;
        }

        if (isset($_GET['message']))
        {
            $URI .= '&message=' . urlencode($_GET['message']);
        }

        if (isset($_GET['messageSuccess']))
        {
            $URI .= '&messageSuccess=' . urlencode($_GET['messageSuccess']);
        }

        /* catsone.com demo domain doesn't relogin. */
        if (strpos(CATSUtility::getIndexName(), '://demo.catsone.com') !== false)
        {
            CATSUtility::transferURL('http://www.catsone.com');
        }
        else
        {
            CATSUtility::transferRelativeURI($URI);
        }
    }
    else if (!ModuleUtility::moduleRequiresAuthentication($_GET['m']))
    {
        /* No authentication required; load the module. */
        ModuleUtility::loadModule($_GET['m']);
    }
    else if (!$_SESSION['CATS']->isLoggedIn())
    {
        /* User isn't logged in and authentication is required; send the user
         * to the login page.
         */
        ModuleUtility::loadModule('login');
    }
    else
    {
        /* Everything's good; load the requested module. */
        $_SESSION['CATS']->logPageView();
        ModuleUtility::loadModule($_GET['m']);
    }
}

if (isset($errorHandler))
{
    $errorHandler->reportErrors();
}

?>
